spec.md

```diff
@@ -0,0 +1,954 @@
+|                                           |
++-------------------------------------------+
+
+Figure 1: CMI client and CMI Plugin together make up
+the machine part of MCM. Both of them are typically expected to run
+as two containers is a same pod. However, both might
+run in different pods as well.
+```
+
+### Machine Lifecycle
+
+```
+  CreateMachine +------------+ DeleteMachine
+ +------------->|  PENDING   +--------------+
+ |              +---+----+---+              |
+ |               VM |    | VM               v
++++              is |    | is              +++
+|X|              Up |    | Down            | |
++-+             +---v----+---+             +-+
+                | AVAILABLE  |
+                +---+----^---+
+                 VM |    | VM
+        is ready to |    | is not ready to
+      schedule pods |    | schedule pods
+                +---v----+---+
+                |   READY    |
+                +------------+
+
+Figure 2: The lifecycle of a dynamically provisioned machine, from
+creation to destruction.
+```
+
+The above diagrams illustrate a general expectation with respect to how a MCM manages the lifecycle of a machine via the API presented in this specification.
+Plugins SHOULD expose all RPCs for an interface: Controller plugins SHOULD implement all RPCs for the `Machine` service.
+Unsupported RPCs SHOULD return an appropriate error code that indicates such (e.g. `CALL_NOT_IMPLEMENTED`).
+
+## Container Machine Interface
+
+This section describes the interface between `CMI-Client` and `CMI-Plugins`.
+
+### RPC Interface
+
+A `CMI-Client` interacts with an `CMI-Plugin` through RPCs.
+Each VM provider MUST provide two sets of RPCs:
+
+* **Identity Service**: These are the set of identity RPCs to be implemented by the CMI-PLugin.
+* **Machine Service**: The CMI-Plugin implement this sets of machine RPCs.
+
+```protobuf
+
+service Identity {
+    rpc GetPluginInfo(GetPluginInfoRequest)
+        returns (GetPluginInfoResponse) {}
+
+    rpc GetPluginCapabilities(GetPluginCapabilitiesRequest)
+        returns (GetPluginCapabilitiesResponse) {}
+
+    rpc Probe (ProbeRequest)
+        returns (ProbeResponse) {}
+}
+
+service Machine {
+    rpc CreateMachine (CreateMachineRequest)
+        returns (CreateMachineResponse) {}
+
+    rpc DeleteMachine (DeleteMachineRequest)
+        returns (DeleteMachineResponse) {}
+
+    rpc GetMachine (GetMachineRequest)
+        returns (GetMachineResponse) {}
+
+    rpc ListMachines (ListMachinesRequest)
+        returns (ListMachinesResponse) {}
+
+    rpc ShutDownMachine (ShutDownMachineRequest)
+        returns (ShutDownMachineResponse) {}
+
+    rpc GetListOfVolumeIDsForExistingPVs(GetListOfVolumeIDsForExistingPVsRequest)
+        returns (GetListOfVolumeIDsForExistingPVsResponse) {}
+}
+
+```
+
+#### Concurrency
+
+In general the CMI-Client is responsible for ensuring that there is no more than one call “in-flight” per machine at a given time.
```

Comment:
I think all of this is needed. At least in our GMP-related implementation it's a very legitimate concern. The CMI plugin should ensure that normally it doesn't make simultaneous calls. However it may "lose state", e.g. crash and be restarted, and issue a parallel call for this reason, so the plugin should in principle be prepared to handle the situation as described. Instead of "call" you may wish to talk about "operation" and include only those operations that change the machine state.